File: packages/types/src/dynamo/report.ts
import type {
  ISODateString,
  JobId,
  OrgId,
  RepoId,
  ReportId,
  RiskLevel,
} from '../common.js';

export type ApprovalStatus = 'PENDING_REVIEW' | 'APPROVED' | 'REJECTED' | 'AUTO_APPROVED';
export type MergeRecommendation =
  | 'APPROVE'
  | 'APPROVE_WITH_FIXES'
  | 'REQUEST_CHANGES'
  | 'BLOCK';

export interface AgentSummaries {
  readonly codeReview?: string;
  readonly infraReview?: string;
  readonly securityReview?: string;
  readonly riskAssessment: string;
}

export interface ReportItem {
  readonly PK: `REPORT#${ReportId}`;
  readonly SK: 'METADATA';
  readonly reportId: ReportId;
  readonly jobId: JobId;
  readonly orgId: OrgId;
  readonly repoId: RepoId;
  readonly riskScore: number;       // 0-100
  readonly riskLevel: RiskLevel;
  readonly mergeRecommendation: MergeRecommendation;
  readonly approvalStatus: ApprovalStatus;
  readonly findingCounts: {
    readonly critical: number;
    readonly high: number;
    readonly medium: number;
    readonly low: number;
    readonly info: number;
    readonly total: number;
  };
  readonly agentSummaries: AgentSummaries;
  readonly reportS3Key?: string;
  readonly githubCheckRunId?: number;
  readonly createdAt: ISODateString;
  readonly updatedAt: ISODateString;
  // GSI1: jobId
  readonly GSI1PK: `JOB#${JobId}`;
  readonly GSI1SK: ISODateString;
  // GSI2: repoId + createdAt
  readonly GSI2PK: `REPO#${RepoId}`;
  readonly GSI2SK: ISODateString;
  // GSI3: orgId#approvalStatus + createdAt
  readonly GSI3PK: `ORG#${OrgId}#${ApprovalStatus}`;
  readonly GSI3SK: ISODateString;
}

export interface Report
  extends Omit<ReportItem, 'PK' | 'SK' | 'GSI1PK' | 'GSI1SK' | 'GSI2PK' | 'GSI2SK' | 'GSI3PK' | 'GSI3SK'> {}

File: libs/finding-schema/src/finding_schema/report.py
from __future__ import annotations

from enum import StrEnum
from typing import Annotated

from pydantic import BaseModel, Field

from .finding import AgentFinding


class RiskLevel(StrEnum):
    CRITICAL = "CRITICAL"
    HIGH = "HIGH"
    MEDIUM = "MEDIUM"
    LOW = "LOW"


class MergeRecommendation(StrEnum):
    APPROVE = "APPROVE"
    APPROVE_WITH_FIXES = "APPROVE_WITH_FIXES"
    REQUEST_CHANGES = "REQUEST_CHANGES"
    BLOCK = "BLOCK"


class FindingCounts(BaseModel):
    model_config = {"frozen": True}

    critical: int = 0
    high: int = 0
    medium: int = 0
    low: int = 0
    info: int = 0

    @property
    def total(self) -> int:
        return self.critical + self.high + self.medium + self.low + self.info


class AgentSummaries(BaseModel):
    model_config = {"frozen": True}

    code_review: str | None = None
    infra_review: str | None = None
    security_review: str | None = None
    risk_assessment: str


class AgentReport(BaseModel):
    """Aggregated output from all agents for a single analysis job."""

    model_config = {"frozen": True}

    job_id: str
    org_id: str
    repo_id: str
    risk_score: Annotated[int, Field(ge=0, le=100)]
    risk_level: RiskLevel
    merge_recommendation: MergeRecommendation
    findings: list[AgentFinding] = Field(default_factory=list)
    finding_counts: FindingCounts
    agent_summaries: AgentSummaries

    @classmethod
    def compute_risk_level(cls, score: int) -> RiskLevel:
        if score >= 80:
            return RiskLevel.CRITICAL
        if score >= 60:
            return RiskLevel.HIGH
        if score >= 40:
            return RiskLevel.MEDIUM
        return RiskLevel.LOW

    @classmethod
    def compute_merge_recommendation(
        cls, risk_level: RiskLevel, has_critical: bool, has_high: bool
    ) -> MergeRecommendation:
        if risk_level == RiskLevel.CRITICAL or has_critical:
            return MergeRecommendation.BLOCK
        if risk_level == RiskLevel.HIGH or has_high:
            return MergeRecommendation.REQUEST_CHANGES
        if risk_level == RiskLevel.MEDIUM:
            return MergeRecommendation.APPROVE_WITH_FIXES
        return MergeRecommendation.APPROVE
